src/N3Index.js

```diff
@@ -0,0 +1,604 @@
+/* eslint-disable no-warning-comments */
+// TODO: Remove the above eslint-disable comment before making a PR.
+import { termToId, termFromId } from './N3DataFactory';
+
+// Determines whether the argument is a string
+function isString(s) {
+  return typeof s === 'string' || s instanceof String;
+}
+
+/**
+ * A three-layered index.
+ */
+export default class N3Index {
+  constructor(config) {
+    // The number of quads is initially zero
+    this._size = 0;
+    // `_graphs` contains subject, predicate, and object indexes per graph
+    this._graphs = Object.create(null);
+    // `_ids` maps entities such as `http://xmlns.com/foaf/0.1/name` to numbers,
+    // saving memory by using only numbers as keys in `_graphs`
+    this._id = 0;
+    this._ids = Object.create(null);
+    this._ids['><'] = 0; // dummy entry, so the first actual key is non-zero
+    this._entities = Object.create(null); // inverse of `_ids`
+    // `_blankNodeIndex` is the index of the last automatically named blank node
+    this._blankNodeIndex = 0;
+
+    this._factory = config.factory;
+  }
+
+  // ## Public properties
+
+  // ### `size` returns the number of quads in the store
+  get size() {
+    // Return the quad count if if was cached
+    let size = this._size;
+    if (size !== null)
+      return size;
+
+    // Calculate the number of quads by counting to the deepest level
+    size = 0;
+    const graphs = this._graphs;
+    let subjects, subject;
+    for (const graphKey in graphs)
+      for (const subjectKey in (subjects = graphs[graphKey].subjects))
+        for (const predicateKey in (subject = subjects[subjectKey]))
+          size += Object.keys(subject[predicateKey]).length;
+    return this._size = size;
+  }
+
+  // ## Public methods
+
+  // ### `addQuad` adds a new quad to the store.
+  // Returns if the quad index has changed, if the quad did not already exist.
+  addQuad(subject, predicate, object, graph) {
+    // TODO: Move this shift into N3Store.
+    // Shift arguments if a quad object is given instead of components
+    if (!predicate)
+      graph = subject.graph, object = subject.object,
+        predicate = subject.predicate, subject = subject.subject;
+
+    // Convert terms to internal string representation
+    subject = termToId(subject);
+    predicate = termToId(predicate);
+    object = termToId(object);
+    graph = termToId(graph);
+
+    // Find the graph that will contain the triple
+    let graphItem = this._graphs[graph];
+    // Create the graph if it doesn't exist yet
+    if (!graphItem) {
+      // TODO: See if converting the following object literal to a TermsIndex class has
+      // any benefit.
+      graphItem = this._graphs[graph] = { subjects: {}, predicates: {}, objects: {} };
```

Comment:
I was thinking about decomposing this class further by introducing a `TermsIndex` class. It would be used like:
```suggestion
      graphItem = this._graphs[graph] = new TermsIndex();
```